rs/cli/src/auth.rs

```diff
@@ -220,3 +199,28 @@ impl Auth {
         }
     }
 }
+
+impl TryFrom<AuthOpts> for Auth {
+    type Error = anyhow::Error;
+    fn try_from(auth_opts: AuthOpts) -> Result<Self, anyhow::Error> {
+        match &auth_opts {
+            AuthOpts {
+                private_key_pem: Some(private_key_pem),
+                hsm_opts: _,
+            } => Ok(Auth::Keyfile {
+                path: private_key_pem.path().to_path_buf(),
+            }),
+            AuthOpts {
+                private_key_pem: _,
+                hsm_opts: Some(hsm_opts),
+            } => Ok(Auth::Hsm {
+                pin: hsm_opts.hsm_pin.clone(),
+                slot: hsm_opts.hsm_slot,
+                key_id: hsm_opts.hsm_key_id.clone(),
+            }),
+            // I think the next line should not fall back to anonymous.
+            // It should always be autodetect and fail if detection fails.
+            _ => Ok(Self::detect_hsm_auth()?.map_or(Auth::Anonymous, |a| a)),
```

Comment:
Not really, because if you want to do some canister interactions, you don't need ic-admin but you need some auth, usually a readonly auth, that is why we have `Anonymous` here. But sometimes you do need some auth for canisters

Reply:
While I won't change this here (I'm leaving it as-is), this definitely needs to be fixed.  In those "some interactions that don't need ic-admin but some auth", a different **type** should be used.  And the program does know this already -- but the circumstance is just not specified in types as it should be.  It should be impossible -- enforced by the type system -- for an action that requires full auth to ever execute with anonymous auth.

Reply:
The real answer is that `ic-admin` is just a wrapper for some canister calls. Meaning that, in theory we don't really need `ic-admin` and that is why we have the difference. Under the hood ic admin has a `ic-canister-client` which has a pool of agents.